guides/building-a-rest-api-spring-boot-vs-micronaut-test-first/micronautframeworkjacksondatabind/java/src/test/java/example/micronaut/SaasSubscriptionJsonTest.java

```diff
@@ -0,0 +1,78 @@
+import io.micronaut.core.io.ResourceLoader;
+import io.micronaut.json.JsonMapper;
+import io.micronaut.test.extensions.junit5.annotation.MicronautTest;
+import org.junit.jupiter.api.Test;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.InputStreamReader;
+import java.io.Reader;
+import java.nio.charset.StandardCharsets;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+@MicronautTest(startApplication = false) // <1>
+class SaasSubscriptionJsonTest {
+
+    @Test
+    void saasSubscriptionSerializationTest(JsonMapper json, ResourceLoader resourceLoader) throws IOException { // <2>
+        SaasSubscription subscription = new SaasSubscription(99L, "Professional", 4900);
+        String expected = getResourceAsString(resourceLoader, "expected.json");
+        String result = json.writeValueAsString(subscription);
+        assertThat(result).isEqualToIgnoringWhitespace(expected);
+        DocumentContext documentContext = JsonPath.parse(result);
+        Number id = documentContext.read("$.id");
+        assertThat(id)
+                .isNotNull()
+                .isEqualTo(99);
+
+        String name = documentContext.read("$.name");
+        assertThat(name)
+                .isNotNull()
+                .isEqualTo("Professional");
+
+        Number cents = documentContext.read("$.cents");
+        assertThat(cents)
+                .isNotNull()
+                .isEqualTo(4900);
+    }
+
+    @Test
+    void saasSubscriptionDeserializationTest(JsonMapper json) throws IOException { // <2>
+        String expected = """
+           {
+               "id":100,
+               "name": "Advanced",
+               "cents":2900
+           }
+           """;
+        assertThat(json.readValue(expected, SaasSubscription.class))
+                .isEqualTo(new SaasSubscription(100L, "Advanced", 2900));
+        assertThat(json.readValue(expected, SaasSubscription.class).id()).isEqualTo(100);
+        assertThat(json.readValue(expected, SaasSubscription.class).name()).isEqualTo("Advanced");
+        assertThat(json.readValue(expected, SaasSubscription.class).cents()).isEqualTo(2900);
+    }
+
+    private static String getResourceAsString(ResourceLoader resourceLoader, String resourceName) {
+        Optional<InputStream> optionalStream = resourceLoader.getResourceAsStream(resourceName);
+        String result = "";
+
+        if (optionalStream.isPresent()) {
+            try (InputStream stream = optionalStream.get();
+                 Reader reader = new InputStreamReader(stream, StandardCharsets.UTF_8);
+                 BufferedReader bufferedReader = new BufferedReader(reader)) {
+                result = bufferedReader.lines().collect(Collectors.joining("\n"));
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        }
+        return result;
+    }
```

Comment:
Is this better? I suspect it doesn't matter that much...  Shame we can't more easily inject file contents into tests from test-resources

```suggestion
    private static String getResourceAsString(ResourceLoader resourceLoader, String resourceName) {
        return resourceLoader.getResourceAsStream(resourceName)
                .flatMap(stream -> {
                    try (InputStream is = stream;
                         Reader reader = new InputStreamReader(is, StandardCharsets.UTF_8);
                         BufferedReader bufferedReader = new BufferedReader(reader)) {
                        return Optional.of(bufferedReader.lines().collect(Collectors.joining("\n")));
                    } catch (IOException e) {
                        e.printStackTrace();
                    }
                    return Optional.empty();
                })
                .orElse("");
    }
```

Reply:
Yes, I should add a feature in Micronaut Test to ease it. Spring boot has some nice quality of live features.